rfcs/2020-08-31-3641-mongo-metrics.md

```diff
@@ -0,0 +1,191 @@
+# RFC 3603 - 2020-08-27 - Collecting metrics from MongoDB
+
+This RFC is to introduce a new metrics source to consume metrics from MongoDB database servers. The high level plan is to implement one source that collects metrics from MongoDB server instances.
+
+Background reading on MongoDB monitoring:
+
+- https://docs.mongodb.com/manual/administration/monitoring/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for MongoDB server metrics.
+
+This RFC will not cover:
+
+- Other databases.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their MongoDB databases are performing.
+
+## Internal Proposal
+
+Build a single source called `mongodb_metrics` (name to be confirmed) to collect MongoDB metrics. We support all non-EOL'ed MongoDB versions.
+
+The recommended implementation is to use the Rust MongoDB client to connect the target database server by the address specified in configuration.
+
+Proposed client:
+
+- https://github.com/mongodb/mongo-rust-driver
+- https://crates.io/crates/mongodb
+
+We'd call the following on the server from the client upon connection:
+
+- [`serverStatus`](https://docs.mongodb.com/manual/reference/command/serverStatus/)
+- [`printShardingStatus`](https://docs.mongodb.com/manual/reference/method/db.printShardingStatus/)
+- [`shardConnPoolStats`](https://docs.mongodb.com/manual/reference/command/shardConnPoolStats/)
+- [`dbStats`](https://docs.mongodb.com/manual/reference/command/dbStats/) on each DB
+- [`collStats`](https://docs.mongodb.com/manual/reference/command/collStats/) per collection
+- [`connPoolStats`](https://docs.mongodb.com/manual/reference/command/connPoolStats/) per pool
+- [`Top`](https://docs.mongodb.com/manual/reference/command/top/)
+- [`indexStats`](https://docs.mongodb.com/manual/reference/operator/aggregation/indexStats/#pipe._S_indexStats)
+- [`replSetGetStatus`](https://docs.mongodb.com/manual/reference/command/replSetGetStatus/)
+
+This should result in the following metrics being collected:
```

Comment:
I don't know much about running mongodb so I didn't audit these especially closely; just looked for inconsistencies. I'm happy to do a closer audit (probably a day or so to get up-to-speed on mongo and figure out what these metrics mean), but this would be a little time intensive so, if we would like that thorough of a review, I think it'd be nice to schedule into the sprint.

Alternatively, I'm happy if they are simply match, or are a superset of, telegraf which I consider to be the most popular tool for this.